source/opt/local_redundancy_elimination.h

```diff
@@ -0,0 +1,45 @@
+// Copyright (c) 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef LIBSPIRV_OPT_LOCAL_REDUNDANCY_ELIMINATION_H_
+#define LIBSPIRV_OPT_LOCAL_REDUNDANCY_ELIMINATION_H_
+
+#include "ir_context.h"
+#include "pass.h"
+#include "value_number_table.h"
+
+namespace spvtools {
+namespace opt {
+
+// See optimizer.hpp for documentation.
+class LocalRedundancyEliminationPass : public Pass {
+ public:
+  const char* name() const override { return "local-redundancy-elimination"; }
+  Status Process(ir::IRContext*) override;
+
+ private:
+  // Deletes instructions in |block| whose value is in |vnTable| or is computed
+  // earlier in |block|.  The values computed in |block| are added to |vnTable|.
```

Comment:
But if it's already in |vnTable|, it's no longer a local CSE.  Or are you thinking of using this same mechanism for global CSE?

Reply:
I wanted this to be general enough to be used for a global version of the same algorithm.  When I start on the global version, we can have a discussion on the exact algorithm, but this function should be reused.